### ISSUE TITLE
Using a device farm for builds?

### ISSUE THREAD

**Issue by mindlapse** [tipsi/tipsi-stripe#505]
Running mobile tests on travis is very problematic:
- We're limited to emulation
- Emulation on containerized environments is limited to ARM based images on android (the x86 based images require hardware acceleration)
- ARM based system images are slow and do not seem to be available (at least in the default repos) for newer versions of android according to the new sdkmanager.

One solution is to use a device farm, where testing occurs on physical devices.  Another option is to use a service like Saucelabs that provides emulators.  If cost wasn't an issue, which device farm / emulation platform would you use?

Saucelabs and AWS Device Farm are two possible options

**Comment by mindlapse** [comment]
Saucelabs may be interesting, since Travis CI integrates with it:
https://docs.travis-ci.com/user/sauce-connect/

**Comment by mindlapse** [comment]
and it appears to be free for open source.  I'm still trying to get the build to pass for Android (although right now there is an issue with Appium connectivity), but I'll start exploring this as an option

**Comment by cybergrind** [comment]
@mindlapse I've tried x86 images it didn't run faster than ARM, so probably it isn't about virtualization but more about TravisCI CPU limitation. And android tests at least green from time to time, so we can play with timeouts here.
But OSX test always fails with a timeout.

I've applied to opensource program on saucelabs.com will see the response. 

AWS device farm isn't a great variant, because project already exists as goodwill from our management and I'm not sure if I really can spend more money on it (we've already spent a lot of money on developers).
I've suggested moving this library to stripe team for maintenance ~ half year ago but they (stripe) responded that they aren't interested in supporting it at that time.

Another option is migration to gitlab and then I will be able to run separate runners on our hardware servers. It isn't perfect (because servers may go away any time in the future) but at least it will have decent performance.

**Comment by fbartho** [comment]
@mindlapse do you have a sense on how long it takes to run a full build + all the tests?

Maybe we could use a CI like Nevercode.io which has a free tier. -- Builds would get queued for simultaneous requests, and there's a timeout, but that might be way cheaper than Saucelabs or AWS Device Farm.

We have a paid Nevercode account, and it can run simulators for UI / Interaction testing.

**Comment by fbartho** [comment]
Another problem with the Device Labs is that it feels simultaneously super overkill for an Open Source project, and also problematic. -- Our company used to use Sauce Labs, and the devices were so flaky that a lot of time was spent debugging unrelated issues.

Instead Tipsi-Stripe seems ideal for a simulator-based testing system. @Github might have a solution in the future?

**Comment by fbartho** [comment]
I know it's super terrible to suggest, but maybe we should consider not having a CI for a while?

It feels like this project is super stuck right now with lots of open and unfinished PRs. We need to make progress for the SCA deadline. Maybe shipping `-beta.#` releases is the better route, so people can test them using NPM and report issues, and then if X-time passes after releasing a beta release without issues filed against it, then we could make a "stable" release then?

**Comment by mindlapse** [comment]
The most recent Travis build (on my fork) ran into a few test related issues and with the timeouts adding up the job was killed after 50 minutes.  If this project gets upgraded to a paid Travis plan then it appears that the build time limit increases to 120 minutes which should be sufficient - I don't expect cost to be an issue with proper support.

For iOS, I see beta->stable approach as a possibly viable solution.  @cybergrind ?

**Comment by mindlapse** [comment]
I'm seeing low resolution (600x320) builds fail on Travis with long build times, due to 'system is not responding' style messages in the emulator (https://i.imgur.com/fW8rw66.png) (https://travis-ci.com/mindlapse/tipsi-stripe/builds/120314334).  These are builds that do not fail on better machines, like the c5n.xlarge instances on Amazon.

I'm currently running a build at 400x300 but it's getting unrealistically low res at that point.  It's likely better for us to shift over to CircleCI, which at least anecdotally has better performance: https://hackernoon.com/why-xod-switched-from-travis-to-circle-ci-98b6832f2d96

**Comment by cybergrind** [comment]
@mindlapse I hope you have an osx machine to run tests. I really don't want to block you with these CI issues.

Probably we would go right now with running tests locally (they still have to pass) but not care about TravisCI until we figure out what to do.

We also can run internal tests on our CI, this will give us a bit more confidence.
And priorities will be: fixing local tests > library upgrades > new features (+ tests).

**Comment by mindlapse** [comment]
@cybergrind, I've managed to get things working on CircleCI with the larger instance types - I'm putting together a PR :)

**Comment by cybergrind** [comment]
@mindlapse ah, that's super!

**Comment by mindlapse** [comment]
I've gathered the CircleCI changes into a commit added to this PR:
https://github.com/tipsi/tipsi-stripe/pull/497

It includes a bunch of test suite changes.  One of the AliPay tests are failing but it doesn't appear to be an issue with the tests, rather with the test page itself.  I've commented that one out for now.  I'm going to start digging into the example app now to pull together some of the payment intent changes
